internal/rotate_est.go

```diff
@@ -0,0 +1,178 @@
+package internal
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"go.mozilla.org/pkcs7"
+)
+
+var (
+	oidKeyUsage         = asn1.ObjectIdentifier([]int{2, 5, 29, 15})
+	oidExtendedKeyUsage = asn1.ObjectIdentifier([]int{2, 5, 29, 37})
+
+	asn1DigitalSignature = []byte{3, 2, 7, 128}
+	asn1TlsWebClientAuth = []byte{48, 10, 6, 8, 43, 6, 1, 5, 5, 7, 3, 2}
+)
+
+type estStep struct{}
+
+func (s estStep) Name() string {
+	return "Generate new certificate"
+}
+
+func (s estStep) Completed(state CertRotationState) bool {
+	return len(state.NewKey) > 0 && len(state.NewCert) > 0
```

Comment:
If a state/step name or index is part of `CertRotationState` context and stored in a file, then there is no need in this method `Completed`.